### PR TITLE
python: missing libxcrypt dep

### PR DESCRIPTION
**Pull request by haampie**
Detected using https://github.com/spack/spack/pull/28109

May need https://github.com/spack/spack/pull/28109

@adamjstewart see also https://peps.python.org/pep-0594/#operating-system-interface, should we just try to disable all of crypt/nis/spwd by default (maybe even without an option to enable it?)

In any case, it would be model Python's dependencies correctly. I'm hoping https://github.com/spack/spack/pull/28109 can be a strict requirement in Gitlab CI to avoid accidental system deps.

### CODE REVIEW

**Comment by adamjstewart** [comment]
I think we should add a `crypt` variant, like we have for all other optional Python modules with external dependencies. Also, according to https://github.com/python/cpython/issues/92238, this dependency was only added in Python 3.9.

**Comment by haampie** [comment]
This whole libcrypt.so.1 thing is the least to say painful to watch. See for example

https://github.com/pypa/manylinux/issues/305

Some distributions no longer have libcrypt.so.1 as it was dropped from glibc, so I think it's good to use our own.

Then `libcrypt.so` has incompatible APIs (and hence a bumped soname), with the flag `--disable-obsolete-api` you get `libcrypt.so.2` iiuc. Maybe the libxcrypt package should be updated too.

Updated libxcrypt. I think we can safely go for libcrypt.so.2; it works fine for Python, and we don't model glibc anyways.

Note that libxcrypt takes 2 seconds to build, so I think adding a variant is not really worth it.

**Comment by adamjstewart** [comment]
I think I would still like a variant, just for consistency with other optional dependencies. We've gotten too many complaints in the past about Spack building a million dependencies. The default will be true, but people can always disable the variant if they really want to speed up build times. Note that you'll also need to add an if-statement to the post-install import tests. I can push a commit for this if you want.

**Comment by haampie** [comment]
Feel free to push to this branch if it's easy 🙃

**Comment by adamjstewart** [comment]
Tested this a bit. Even on Python 3.8, it's still linking to crypt. I think the issue I linked to was mistaken and Python 3.9 just improved pkgconfig detection. Left the dep in for all versions < 3.13 just to be safe.

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -370,6 +370,9 @@ class Python(Package):
     variant("tkinter", default=False, description="Build tkinter module")
     variant("uuid", default=True, description="Build uuid module")
     variant("tix", default=False, description="Build Tix module")
+    variant("crypt", default=True, description="Build crypt module", when="@:3.12 platform=linux")
+    variant("crypt", default=True, description="Build crypt module", when="@:3.12 platform=darwin")
+    variant("crypt", default=True, description="Build crypt module", when="@:3.12 platform=cray")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py:375]: @becker33 excited to see cray being removed as a platform, but would love a `platform=unix:` option or a "not" operator in our spec syntax.

**Review by adamjstewart — APPROVED** [review]

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/python/package.py [hunk 2]
```diff
@@ -560,6 +564,14 @@ def determine_variants(cls, exes, version_str):
             except ProcessError:
                 variants += "~tix"
 
+        # Some modules are platform-dependent
+        if not self.spec.satisfies("platform=windows"):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/python/package.py:568]: This breaks detection, since it's inside a class method (`self` doesn't exist here).

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/python/package.py:568]: Oops, will fix using `is_windows`

**Pull request merged**